macros/support/src/class/class.rs

```diff
@@ -0,0 +1,308 @@
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let mut class = None;
+        let mut methods: HashMap<syn::Path, Vec<syn::ImplItemMethod>> = HashMap::new();
+        while !input.is_empty() {
+            let docs = input.call(syn::Attribute::parse_outer)?;
+            if let Some(a) = docs.iter().find(|a| !a.path.is_ident("doc")) {
+                return Err(syn::Error::new(a.path.span(), "Unrecognized attribute"));
+            }
+
+            if !input.peek(syn::Token!(impl)) {
+                class = Some(Self::parse_class(input, docs)?);
+            } else {
+                let item = input.parse::<syn::ItemImpl>()?;
+                // TODO: ensure that class idents line up
+                let interface = match item.trait_ {
+                    Some((_, i, _)) => i,
+                    None => {
+                        return Err(syn::Error::new(
+                            item.span(),
+                            "Impl must be for an interface",
+                        ))
+                    }
+                };
+                let ms = item
+                    .items
+                    .into_iter()
+                    .map(|i| match i {
+                        syn::ImplItem::Method(m) => Ok(m),
+                        _ => Err(syn::Error::new(
+                            i.span().clone(),
+                            "only trait methods are allowed when implementing an interface",
+                        )),
+                    })
+                    .collect::<syn::Result<Vec<_>>>()?;
+
+                if let Some(_) = methods.insert(interface.clone(), ms) {
+                    return Err(syn::Error::new(
+                        interface.span().clone(),
+                        "interface was redefined",
+                    ));
+                }
+            }
+        }
+        let mut class = match class {
+            Some(c) => c,
+            None => {
+                return Err(syn::Error::new(
+                    input.span().clone(),
+                    "no class was defined",
+                ));
+            }
+        };
+        class.methods = methods;
+        Ok(class)
+    }
+}
+
+mod keywords {
+    syn::custom_keyword!(class);
+    syn::custom_keyword!(factory);
+}
+
+pub struct Interface {
+    pub path: syn::Path,
+    pub parent: Option<Box<Interface>>,
+}
+
+impl Interface {
+    /// Creates an intialized VTable for the interface
+    pub fn to_initialized_vtable_tokens(&self, class: &Class, offset: usize) -> TokenStream {
+        let class_name = &class.name;
+        let vtable_ident = self.vtable_ident();
+        let vtable_type = self.to_vtable_type_tokens();
+        let parent = match self.parent.as_ref() {
+            Some(p) => p.to_initialized_vtable_tokens(class, offset),
+            None => Self::iunknown_tokens(class, offset),
+        };
+        let fields = class.methods.get(&self.path).unwrap().iter().map(|m| {
```

Comment:
I added error checking in the parsing phase that fixes this. We should refactor the code to make this unwrap unnecessary but for now it _shouldn't_ ever panic 🤞 .

Reply:
Neat, the error is on the right span as well :ok_hand:

Reply:
FWIW you might want to include the name of the missing path as well, so that they make a bit more sense in editors (VSCode "problems" view) that only display the error and line+column. Similar to `error[E0412]: cannot find type `TheMissingImplementation` in this scope`